modules/datastore/index.js

```diff
@@ -149,9 +155,124 @@ function getDatum(ctx) {
+      switch (columnsResult.length) {
+        case 0:
+          return columnDefinitions;
+        case 3:
+          return [];
+        default:
+          return getMissingColumns(columnsResult, columnDefinitions);
+      }
+    })
+};
+
+/**
+ * Creates the specified columns on the table
+ * @param {Object} ctx
+ * @param {String} tableName
+ * @param {Array<Object>} columnDefinitions
+ */
+const _createTableColumns = (ctx, appSlug, tableId, tableName, columnDefinitions) => {
+  return getColumnsToCreate(ctx, appSlug, tableId, tableName, columnDefinitions)
+    .then((columns) => {
+      // This really shouldn't ever happen, but if it does
+      // we don't want to error out trying to create columns
+      // that already exist
+      if (columns.length === 0) {
+        return tableId;
+      }
+
+      var tableColumnQueries = columns.map((columnData) => {
+        columnData.input.tableId = tableId;
+        return ctx.graphql.query(CREATE_TABLE_COLUMN_QUERY, columnData);
+      });
+
+      return Promise.all(tableColumnQueries)
+        .then((results) => {
+          if (results.some(resultContainsErrorIgnorePsycoPGDuplicate)) {
+            return Promise.reject('Error creating table columns');
+          }
+          return tableId;
+        });
+    });
+};
+
+function _createTable(ctx) {
+  return function (appSlug, tableName, tableDescription, columnDefinitions) {
+    const createInput = {
+      input: {
+        applicationId: "Application:"+appSlug,
+        name: tableName,
+        description: tableDescription
+      }
+    };
+    return ctx.graphql.query(CREATE_TABLE_QUERY, createInput)
+      .then((result) => {
+        if (resultContainsError(result)) {
+          return Promise.reject(result.errors[0]);
+        }
+
+        return result.data.createTable.table.id;
+      })
+      .then((tableId) => {
+        return _createTableColumns(ctx, appSlug, tableId, tableName, columnDefinitions);
+    });
+  };
+};
+
+function _findTable(ctx) {
+  return function(tableName, appSlug) {
+    return ctx.graphql.query(createFindTableQuery(), createFindTableQueryVariables(tableName, appSlug))
+      .then((result) => {
+        if (resultContainsError(result)) {
+          return Promise.reject(result.errors[0]);
+        }
+        return result.data.node.table.id
+      });
   }
 }
 
+function _ensure(ctx) {
+  return function(appSlug, tableName, tableDescription, columnDefinitions) {
+    return ctx.graphql.query(createFindTableQuery(), createFindTableQueryVariables(tableName, appSlug))
+      .then((result) => {
+        if (resultContainsError(result)) {
+          if (result.data.node.table === null) {
+            return _createTable(ctx)(appSlug, tableName, tableDescription, columnDefinitions);
+          } else {
+            return Promise.reject(result.errors[0]);
+          }
+        }
+        return _createTableColumns(ctx, appSlug, result.data.node.table.id, tableName, columnDefinitions );
```

Comment:
i changed the logic here to create columns every time, and ignore psycopg2 duplicate errors
and only catch real errors. this will break however if we start catching the psycopg2 error in drone_web

Reply:
👍